File: src/main/java/de/tum/cit/aet/artemis/exercise/dto/ExerciseDetailsResponseDTO.java
package de.tum.cit.aet.artemis.exercise.dto;

import java.time.ZonedDateTime;
import java.util.List;
import java.util.Objects;
import java.util.Set;
import java.util.stream.Collectors;

import org.hibernate.Hibernate;
import org.jspecify.annotations.Nullable;

import com.fasterxml.jackson.annotation.JsonInclude;

import de.tum.cit.aet.artemis.assessment.domain.AssessmentType;
import de.tum.cit.aet.artemis.exercise.domain.DifficultyLevel;
import de.tum.cit.aet.artemis.exercise.domain.Exercise;
import de.tum.cit.aet.artemis.exercise.domain.ExerciseMode;
import de.tum.cit.aet.artemis.exercise.domain.ExerciseType;
import de.tum.cit.aet.artemis.exercise.domain.IncludedInOverallScore;
import de.tum.cit.aet.artemis.fileupload.domain.FileUploadExercise;
import de.tum.cit.aet.artemis.modeling.domain.DiagramType;
import de.tum.cit.aet.artemis.modeling.domain.ModelingExercise;
import de.tum.cit.aet.artemis.programming.domain.ProgrammingExercise;
import de.tum.cit.aet.artemis.programming.domain.submissionpolicy.SubmissionPolicy;
import de.tum.cit.aet.artemis.quiz.domain.QuizExercise;
import de.tum.cit.aet.artemis.quiz.domain.QuizMode;
import de.tum.cit.aet.artemis.quiz.dto.QuizBatchDTO;
import de.tum.cit.aet.artemis.text.domain.TextExercise;

/**
 * Flat, discriminator-based DTO containing the shared and subtype-specific fields used by the student exercise details view.
 *
 * @param id                                         the exercise identifier
 * @param type                                       the JSON exercise discriminator
 * @param exerciseType                               the exercise type enum
 * @param title                                      the title, if available
 * @param shortName                                  the short name, if available
 * @param problemStatement                           the problem statement, if available
 * @param gradingInstructions                        the grading instructions, if visible
 * @param releaseDate                                the release date, if configured
 * @param startDate                                  the start date, if configured
 * @param dueDate                                    the due date, if configured
 * @param assessmentDueDate                          the assessment due date, if configured
 * @param exampleSolutionPublicationDate             the example-solution publication date, if configured
 * @param maxPoints                                  the maximum points, if configured
 * @param bonusPoints                                the bonus points, if configured
 * @param assessmentType                             the assessment type, if configured
 * @param difficulty                                 the difficulty, if configured
 * @param mode                                       the participation mode
 * @param includedInOverallScore                     how the exercise contributes to the course score
 * @param allowComplaintsForAutomaticAssessments     whether automatic-assessment complaints are allowed
 * @param allowFeedbackRequests                      whether feedback requests are allowed
 * @param presentationScoreEnabled                   whether presentation scores are enabled, if configured
 * @param secondCorrectionEnabled                    whether second correction is enabled
 * @param feedbackSuggestionModule                   the feedback suggestion module, if configured
 * @param categories                                 the initialized serialized categories, or absent when not loaded
 * @param teamMode                                   whether the exercise uses team mode
 * @param studentAssignedTeamId                      the current student's assigned team identifier, if any
 * @param studentAssignedTeamIdComputed              whether the assigned team identifier was computed
 * @param course                                     the narrow course context, if initialized
 * @param studentParticipations                      the initialized student participations and filtered histories, or absent when not loaded
 * @param allowOnlineEditor                          whether the online editor is allowed for programming exercises
 * @param allowOfflineIde                            whether offline IDE use is allowed for programming exercises
 * @param allowOnlineIde                             whether the online IDE is allowed for programming exercises
 * @param staticCodeAnalysisEnabled                  whether static code analysis is enabled for programming exercises
 * @param showTestNamesToStudents                    whether test names are shown for programming exercises
 * @param buildAndTestStudentSubmissionsAfterDueDate the post-due-date build time for programming exercises, if configured
 * @param releaseTestsWithExampleSolution            whether tests are released with the example solution for programming exercises
 * @param submissionPolicy                           the initialized programming submission policy, if configured
 * @param visibleToStudents                          whether a quiz is visible to students
 * @param randomizeQuestionOrder                     whether quiz question order is randomized
 * @param allowedNumberOfAttempts                    the allowed quiz attempts, if configured
 * @param remainingNumberOfAttempts                  the current student's remaining quiz attempts, if available
 * @param quizMode                                   the quiz mode, if applicable
 * @param duration                                   the quiz duration in seconds, if applicable
 * @param quizBatches                                the initialized, server-selected quiz batches, or absent when not loaded
 * @param quizStarted                                whether the quiz has started
 * @param quizEnded                                  whether the quiz has ended
 * @param exampleSolution                            the filtered text or file-upload example solution, if visible
 * @param filePattern                                the accepted file pattern for file-upload exercises, if applicable
 * @param diagramType                                the modeling diagram type, if applicable
 * @param exampleSolutionModel                       the filtered modeling example-solution model, if visible
 * @param exampleSolutionExplanation                 the filtered modeling example-solution explanation, if visible
 */
@JsonInclude(JsonInclude.Include.NON_EMPTY)
public record ExerciseDetailsResponseDTO(Long id, String type, ExerciseType exerciseType, @Nullable String title, @Nullable String shortName, @Nullable String problemStatement,
        @Nullable String gradingInstructions, @Nullable ZonedDateTime releaseDate, @Nullable ZonedDateTime startDate, @Nullable ZonedDateTime dueDate,
        @Nullable ZonedDateTime assessmentDueDate, @Nullable ZonedDateTime exampleSolutionPublicationDate, @Nullable Double maxPoints, @Nullable Double bonusPoints,
        @Nullable AssessmentType assessmentType, @Nullable DifficultyLevel difficulty, ExerciseMode mode, IncludedInOverallScore includedInOverallScore,
        boolean allowComplaintsForAutomaticAssessments, boolean allowFeedbackRequests, @Nullable Boolean presentationScoreEnabled, boolean secondCorrectionEnabled,
        @Nullable String feedbackSuggestionModule, @Nullable Set<String> categories, boolean teamMode, @Nullable Long studentAssignedTeamId, boolean studentAssignedTeamIdComputed,
        @Nullable ExerciseDetailsCourseDTO course, @Nullable List<ExerciseDetailsParticipationDTO> studentParticipations, @Nullable Boolean allowOnlineEditor,
        @Nullable Boolean allowOfflineIde, @Nullable Boolean allowOnlineIde, @Nullable Boolean staticCodeAnalysisEnabled, @Nullable Boolean showTestNamesToStudents,
        @Nullable ZonedDateTime buildAndTestStudentSubmissionsAfterDueDate, @Nullable Boolean releaseTestsWithExampleSolution,
        @Nullable ExerciseDetailsSubmissionPolicyDTO submissionPolicy, @Nullable Boolean visibleToStudents, @Nullable Boolean randomizeQuestionOrder,
        @Nullable Integer allowedNumberOfAttempts, @Nullable Integer remainingNumberOfAttempts, @Nullable QuizMode quizMode, @Nullable Integer duration,
        @Nullable Set<QuizBatchDTO> quizBatches, @Nullable Boolean quizStarted, @Nullable Boolean quizEnded, @Nullable String exampleSolution, @Nullable String filePattern,
        @Nullable DiagramType diagramType, @Nullable String exampleSolutionModel, @Nullable String exampleSolutionExplanation) {

    /**
     * Maps an authorized and filtered exercise to the explicit student details contract without initializing lazy associations.
     *
     * @param exercise the authorized and filtered exercise
     * @return the flat exercise details DTO
     */
    public static ExerciseDetailsResponseDTO of(Exercise exercise) {
        Objects.requireNonNull(exercise, "The exercise must be set");

        Set<String> categories = exercise.getCategories() != null && Hibernate.isInitialized(exercise.getCategories()) ? Set.copyOf(exercise.getCategories()) : null;
        List<ExerciseDetailsParticipationDTO> studentParticipations = exercise.getStudentParticipations() != null && Hibernate.isInitialized(exercise.getStudentParticipations())
                ? exercise.getStudentParticipations().stream().filter(Objects::nonNull).map(ExerciseDetailsParticipationDTO::of).toList()
                : null;
        var courseEntity = exercise.getCourseViaExerciseGroupOrCourseMember();
        ExerciseDetailsCourseDTO course = courseEntity != null && Hibernate.isInitialized(courseEntity) ? ExerciseDetailsCourseDTO.of(courseEntity) : null;

        Boolean allowOnlineEditor = null;
        Boolean allowOfflineIde = null;
        Boolean allowOnlineIde = null;
        Boolean staticCodeAnalysisEnabled = null;
        Boolean showTestNamesToStudents = null;
        ZonedDateTime buildAndTestStudentSubmissionsAfterDueDate = null;
        Boolean releaseTestsWithExampleSolution = null;
        ExerciseDetailsSubmissionPolicyDTO submissionPolicy = null;
        Boolean visibleToStudents = null;
        Boolean randomizeQuestionOrder = null;
        Integer allowedNumberOfAttempts = null;
        Integer remainingNumberOfAttempts = null;
        QuizMode quizMode = null;
        Integer duration = null;
        Set<QuizBatchDTO> quizBatches = null;
        Boolean quizStarted = null;
        Boolean quizEnded = null;
        String exampleSolution = null;
        String filePattern = null;
        DiagramType diagramType = null;
        String exampleSolutionModel = null;
        String exampleSolutionExplanation = null;

        switch (exercise) {
            case ProgrammingExercise programmingExercise -> {
                allowOnlineEditor = programmingExercise.isAllowOnlineEditor();
                allowOfflineIde = programmingExercise.isAllowOfflineIde();
                allowOnlineIde = programmingExercise.isAllowOnlineIde();
                staticCodeAnalysisEnabled = programmingExercise.isStaticCodeAnalysisEnabled();
                showTestNamesToStudents = programmingExercise.getShowTestNamesToStudents();
                buildAndTestStudentSubmissionsAfterDueDate = programmingExercise.getBuildAndTestStudentSubmissionsAfterDueDate();
                releaseTestsWithExampleSolution = programmingExercise.isReleaseTestsWithExampleSolution();
                SubmissionPolicy policy = programmingExercise.getSubmissionPolicy();
                if (policy != null && Hibernate.isInitialized(policy)) {
                    submissionPolicy = ExerciseDetailsSubmissionPolicyDTO.of(policy);
                }
            }
            case QuizExercise quizExercise -> {
                visibleToStudents = quizExercise.isVisibleToStudents();
                randomizeQuestionOrder = quizExercise.isRandomizeQuestionOrder();
                allowedNumberOfAttempts = quizExercise.getAllowedNumberOfAttempts();
                remainingNumberOfAttempts = quizExercise.getRemainingNumberOfAttempts();
                quizMode = quizExercise.getQuizMode();
                duration = quizExercise.getDuration();
                if (quizExercise.getQuizBatches() != null && Hibernate.isInitialized(quizExercise.getQuizBatches())) {
                    quizBatches = quizExercise.getQuizBatches().stream().map(QuizBatchDTO::of).collect(Collectors.toUnmodifiableSet());
                }
                quizStarted = quizExercise.isQuizStarted();
                quizEnded = quizExercise.isQuizEnded();
            }
            case TextExercise textExercise -> exampleSolution = textExercise.getExampleSolution();
            case ModelingExercise modelingExercise -> {
                diagramType = modelingExercise.getDiagramType();
                exampleSolutionModel = modelingExercise.getExampleSolutionModel();
                exampleSolutionExplanation = modelingExercise.getExampleSolutionExplanation();
            }
            case FileUploadExercise fileUploadExercise -> {
                exampleSolution = fileUploadExercise.getExampleSolution();
                filePattern = fileUploadExercise.getFilePattern();
            }
            default -> throw new IllegalArgumentException("Unsupported exercise type: " + exercise.getClass().getName());
        }

        return new ExerciseDetailsResponseDTO(exercise.getId(), exercise.getType(), exercise.getExerciseType(), exercise.getTitle(), exercise.getShortName(),
                exercise.getProblemStatement(), exercise.getGradingInstructions(), exercise.getReleaseDate(), exercise.getStartDate(), exercise.getDueDate(),
                exercise.getAssessmentDueDate(), exercise.getExampleSolutionPublicationDate(), exercise.getMaxPoints(), exercise.getBonusPoints(), exercise.getAssessmentType(),
                exercise.getDifficulty(), exercise.getMode(), exercise.getIncludedInOverallScore(), exercise.getAllowComplaintsForAutomaticAssessments(),
                exercise.getAllowFeedbackRequests(), exercise.getPresentationScoreEnabled(), exercise.getSecondCorrectionEnabled(), exercise.getFeedbackSuggestionModule(),
                categories, exercise.isTeamMode(), exercise.getStudentAssignedTeamId(), exercise.isStudentAssignedTeamIdComputed(), course, studentParticipations,
                allowOnlineEditor, allowOfflineIde, allowOnlineIde, staticCodeAnalysisEnabled, showTestNamesToStudents, buildAndTestStudentSubmissionsAfterDueDate,
                releaseTestsWithExampleSolution, submissionPolicy, visibleToStudents, randomizeQuestionOrder, allowedNumberOfAttempts, remainingNumberOfAttempts, quizMode,
                duration, quizBatches, quizStarted, quizEnded, exampleSolution, filePattern, diagramType, exampleSolutionModel, exampleSolutionExplanation);
    }
}

File: src/main/webapp/app/exercise/shared/entities/exercise/exercise-details-response.dto.ts
import { User, UserPublicInfoDTO } from 'app/account/user/user.model';
import { AssessmentType } from 'app/assessment/shared/entities/assessment-type.model';
import { Course, CourseInformationSharingConfiguration } from 'app/course/shared/entities/course.model';
import { ExerciseCategory, SerializedExerciseCategory } from 'app/exercise/shared/entities/exercise/exercise-category.model';
import { DifficultyLevel, Exercise, ExerciseMode, ExerciseType, IncludedInOverallScore } from 'app/exercise/shared/entities/exercise/exercise.model';
import { InitializationState, ParticipationType } from 'app/exercise/shared/entities/participation/participation.model';
import { ProgrammingExerciseStudentParticipation } from 'app/exercise/shared/entities/participation/programming-exercise-student-participation.model';
import { StudentParticipation } from 'app/exercise/shared/entities/participation/student-participation.model';
import {
    LockRepositoryPolicy,
    SubmissionPenaltyPolicy,
    SubmissionPolicy,
    SubmissionPolicyType,
} from 'app/exercise/shared/entities/submission/submission-policy.model';
import { SubmissionResponseDTO, fromSubmissionResponseDTO } from 'app/exercise/shared/entities/submission/submission-response.dto';
import { Team } from 'app/exercise/shared/entities/team/team.model';
import { convertDateStringFromServer } from 'app/foundation/util/date.utils';
import { parseJson } from 'app/foundation/util/json.util';
import { FileUploadExercise } from 'app/fileupload/shared/entities/file-upload-exercise.model';
import { ModelingExercise } from 'app/modeling/shared/entities/modeling-exercise.model';
import { ProgrammingExercise } from 'app/programming/shared/entities/programming-exercise.model';
import { QuizBatch, QuizExercise, QuizMode } from 'app/quiz/shared/entities/quiz-exercise.model';
import { TextExercise } from 'app/text/shared/entities/text-exercise.model';
import { UMLDiagramType } from '@tumaet/apollon';

export interface ExerciseDetailsCourseDTO {
    id: number;
    title?: string;
    shortName?: string;
    studentGroupName?: string;
    teachingAssistantGroupName?: string;
    editorGroupName?: string;
    instructorGroupName?: string;
    accuracyOfScores?: number;
    complaintsEnabled?: boolean;
    requestMoreFeedbackEnabled?: boolean;
    courseInformationSharingConfiguration?: CourseInformationSharingConfiguration;
}

export interface ExerciseDetailsSubmissionPolicyDTO {
    id: number;
    type: SubmissionPolicyType.LOCK_REPOSITORY | SubmissionPolicyType.SUBMISSION_PENALTY;
    submissionLimit?: number;
    active?: boolean;
    exceedingPenalty?: number;
}

export interface ExerciseDetailsTeamDTO {
    id: number;
    name?: string;
    shortName?: string;
    students?: UserPublicInfoDTO[];
}

export interface ExerciseDetailsParticipationDTO {
    id: number;
    type: ParticipationType.STUDENT | ParticipationType.PROGRAMMING;
    testRun: boolean;
    initializationState?: InitializationState;
    initializationDate?: string;
    individualDueDate?: string;
    presentationScore?: number;
    student?: UserPublicInfoDTO;
    team?: ExerciseDetailsTeamDTO;
    submissions?: SubmissionResponseDTO[];
}

export interface ExerciseDetailsQuizBatchDTO {
    id: number;
    startTime?: string;
    started?: boolean;
    ended?: boolean;
}

export interface ExerciseDetailsResponseDTO {
    id: number;
    type: ExerciseType;
    exerciseType: ExerciseType;
    title?: string;
    shortName?: string;
    problemStatement?: string;
    gradingInstructions?: string;
    releaseDate?: string;
    startDate?: string;
    dueDate?: string;
    assessmentDueDate?: string;
    exampleSolutionPublicationDate?: string;
    maxPoints?: number;
    bonusPoints?: number;
    assessmentType?: AssessmentType;
    difficulty?: DifficultyLevel;
    mode: ExerciseMode;
    includedInOverallScore: IncludedInOverallScore;
    allowComplaintsForAutomaticAssessments: boolean;
    allowFeedbackRequests: boolean;
    presentationScoreEnabled?: boolean;
    secondCorrectionEnabled: boolean;
    feedbackSuggestionModule?: string;
    categories?: string[];
    teamMode: boolean;
    studentAssignedTeamId?: number;
    studentAssignedTeamIdComputed: boolean;
    course?: ExerciseDetailsCourseDTO;
    studentParticipations?: ExerciseDetailsParticipationDTO[];
    allowOnlineEditor?: boolean;
    allowOfflineIde?: boolean;
    allowOnlineIde?: boolean;
    staticCodeAnalysisEnabled?: boolean;
    showTestNamesToStudents?: boolean;
    buildAndTestStudentSubmissionsAfterDueDate?: string;
    releaseTestsWithExampleSolution?: boolean;
    submissionPolicy?: ExerciseDetailsSubmissionPolicyDTO;
    visibleToStudents?: boolean;
    randomizeQuestionOrder?: boolean;
    allowedNumberOfAttempts?: number;
    remainingNumberOfAttempts?: number;
    quizMode?: QuizMode;
    duration?: number;
    quizBatches?: ExerciseDetailsQuizBatchDTO[];
    quizStarted?: boolean;
    quizEnded?: boolean;
    exampleSolution?: string;
    filePattern?: string;
    diagramType?: UMLDiagramType;
    exampleSolutionModel?: string;
    exampleSolutionExplanation?: string;
}

/**
 * Converts the explicit exercise-details wire DTO into the existing component-facing exercise subtype and reconnects its relationships.
 */
export function fromExerciseDetailsResponseDTO(dto: ExerciseDetailsResponseDTO): Exercise {
    const course = dto.course ? fromExerciseDetailsCourseDTO(dto.course) : undefined;
    const exercise = createExercise(dto, course);

    exercise.id = dto.id;
    exercise.title = dto.title;
    exercise.shortName = dto.shortName;
    exercise.problemStatement = dto.problemStatement;
    exercise.gradingInstructions = dto.gradingInstructions;
    exercise.releaseDate = convertDateStringFromServer(dto.releaseDate);
    exercise.startDate = convertDateStringFromServer(dto.startDate);
    exercise.dueDate = convertDateStringFromServer(dto.dueDate);
    exercise.assessmentDueDate = convertDateStringFromServer(dto.assessmentDueDate);
    exercise.exampleSolutionPublicationDate = convertDateStringFromServer(dto.exampleSolutionPublicationDate);
    exercise.maxPoints = dto.maxPoints;
    exercise.bonusPoints = dto.bonusPoints;
    exercise.assessmentType = dto.assessmentType;
    exercise.difficulty = dto.difficulty;
    exercise.mode = dto.mode;
    exercise.includedInOverallScore = dto.includedInOverallScore;
    exercise.allowComplaintsForAutomaticAssessments = dto.allowComplaintsForAutomaticAssessments;
    exercise.allowFeedbackRequests = dto.allowFeedbackRequests;
    exercise.presentationScoreEnabled = dto.presentationScoreEnabled;
    exercise.secondCorrectionEnabled = dto.secondCorrectionEnabled;
    exercise.feedbackSuggestionModule = dto.feedbackSuggestionModule;
    exercise.categories = dto.categories?.map(fromSerializedCategory).filter((category): category is ExerciseCategory => category !== undefined);
    exercise.teamMode = dto.teamMode;
    exercise.studentAssignedTeamId = dto.studentAssignedTeamId;
    exercise.studentAssignedTeamIdComputed = dto.studentAssignedTeamIdComputed;
    exercise.course = course;

    if (exercise instanceof ProgrammingExercise) {
        exercise.allowOnlineEditor = dto.allowOnlineEditor;
        exercise.allowOfflineIde = dto.allowOfflineIde;
        exercise.allowOnlineIde = dto.allowOnlineIde;
        exercise.staticCodeAnalysisEnabled = dto.staticCodeAnalysisEnabled;
        exercise.showTestNamesToStudents = dto.showTestNamesToStudents;
        exercise.buildAndTestStudentSubmissionsAfterDueDate = convertDateStringFromServer(dto.buildAndTestStudentSubmissionsAfterDueDate);
        exercise.releaseTestsWithExampleSolution = dto.releaseTestsWithExampleSolution;
        exercise.submissionPolicy = dto.submissionPolicy ? fromExerciseDetailsSubmissionPolicyDTO(dto.submissionPolicy) : undefined;
    } else if (exercise instanceof QuizExercise) {
        exercise.visibleToStudents = dto.visibleToStudents;
        exercise.randomizeQuestionOrder = dto.randomizeQuestionOrder;
        exercise.allowedNumberOfAttempts = dto.allowedNumberOfAttempts;
        exercise.remainingNumberOfAttempts = dto.remainingNumberOfAttempts;
        exercise.quizMode = dto.quizMode;
        exercise.duration = dto.duration;
        exercise.quizBatches = dto.quizBatches?.map(fromExerciseDetailsQuizBatchDTO);
        exercise.quizStarted = dto.quizStarted;
        exercise.quizEnded = dto.quizEnded;
    } else if (exercise instanceof TextExercise) {
        exercise.exampleSolution = dto.exampleSolution;
    } else if (exercise instanceof FileUploadExercise) {
        exercise.exampleSolution = dto.exampleSolution;
        exercise.filePattern = dto.filePattern;
    } else if (exercise instanceof ModelingExercise) {
        exercise.exampleSolutionModel = dto.exampleSolutionModel;
        exercise.exampleSolutionExplanation = dto.exampleSolutionExplanation;
    }

    exercise.studentParticipations = dto.studentParticipations?.map((participation) => fromExerciseDetailsParticipationDTO(participation, exercise));
    return exercise;
}

function createExercise(dto: ExerciseDetailsResponseDTO, course: Course | undefined): Exercise {
    switch (dto.type) {
        case ExerciseType.PROGRAMMING:
            return new ProgrammingExercise(course, undefined);
        case ExerciseType.QUIZ:
            return new QuizExercise(course, undefined);
        case ExerciseType.TEXT:
            return new TextExercise(course, undefined);
        case ExerciseType.MODELING:
            return new ModelingExercise(dto.diagramType!, course, undefined);
        case ExerciseType.FILE_UPLOAD:
            return new FileUploadExercise(course, undefined);
    }
}

function fromExerciseDetailsCourseDTO(dto: ExerciseDetailsCourseDTO): Course {
    const course = new Course();
    course.id = dto.id;
    course.title = dto.title;
    course.shortName = dto.shortName;
    course.studentGroupName = dto.studentGroupName;
    course.teachingAssistantGroupName = dto.teachingAssistantGroupName;
    course.editorGroupName = dto.editorGroupName;
    course.instructorGroupName = dto.instructorGroupName;
    course.accuracyOfScores = dto.accuracyOfScores;
    course.complaintsEnabled = dto.complaintsEnabled;
    course.requestMoreFeedbackEnabled = dto.requestMoreFeedbackEnabled;
    course.courseInformationSharingConfiguration = dto.courseInformationSharingConfiguration;
    return course;
}

function fromExerciseDetailsParticipationDTO(dto: ExerciseDetailsParticipationDTO, exercise: Exercise): StudentParticipation {
    const participation = dto.type === ParticipationType.PROGRAMMING ? new ProgrammingExerciseStudentParticipation() : new StudentParticipation();
    participation.id = dto.id;
    participation.testRun = dto.testRun;
    participation.initializationState = dto.initializationState;
    participation.initializationDate = convertDateStringFromServer(dto.initializationDate);
    participation.individualDueDate = convertDateStringFromServer(dto.individualDueDate);
    participation.presentationScore = dto.presentationScore;
    participation.student = dto.student ? Object.assign(new User(), dto.student) : undefined;
    participation.team = dto.team ? fromExerciseDetailsTeamDTO(dto.team) : undefined;
    participation.exercise = exercise;
    participation.submissions = dto.submissions?.map(fromSubmissionResponseDTO);
    participation.submissions?.forEach((submission) => (submission.participation = participation));
    return participation;
}

function fromExerciseDetailsTeamDTO(dto: ExerciseDetailsTeamDTO): Team {
    const team = new Team();
    team.id = dto.id;
    team.name = dto.name;
    team.shortName = dto.shortName;
    team.students = dto.students?.map((student) => Object.assign(new User(), student));
    return team;
}

function fromExerciseDetailsSubmissionPolicyDTO(dto: ExerciseDetailsSubmissionPolicyDTO): SubmissionPolicy {
    const policy = dto.type === SubmissionPolicyType.LOCK_REPOSITORY ? new LockRepositoryPolicy() : new SubmissionPenaltyPolicy();
    policy.id = dto.id;
    policy.type = dto.type;
    policy.submissionLimit = dto.submissionLimit;
    policy.active = dto.active;
    policy.exceedingPenalty = dto.exceedingPenalty;
    return policy;
}

function fromExerciseDetailsQuizBatchDTO(dto: ExerciseDetailsQuizBatchDTO): QuizBatch {
    const quizBatch = new QuizBatch();
    quizBatch.id = dto.id;
    quizBatch.startTime = convertDateStringFromServer(dto.startTime);
    quizBatch.started = dto.started;
    quizBatch.ended = dto.ended;
    return quizBatch;
}

function fromSerializedCategory(category: string): ExerciseCategory | undefined {
    try {
        const parsedCategory = parseJson<SerializedExerciseCategory>(category);
        return new ExerciseCategory(parsedCategory.category, parsedCategory.color);
    } catch {
        return undefined;
    }
}
